File: python/agentctl/verifier.py
"""Reference fail-closed verifier for Agent Action Assertion v1."""

from __future__ import annotations

import uuid
from dataclasses import dataclass
from datetime import datetime, timezone
from typing import Any

from cryptography.hazmat.primitives.asymmetric.ed25519 import Ed25519PublicKey

from .audit import AuditEvent, AuditSink
from .models import AuthorizationEvidence, PrincipalRecord, RequestContext
from .protocol import (
    MAX_TTL_SECONDS,
    AssertionErrorCode,
    canonical_request_target,
    normalize_content_type,
    parse_assertion,
    sha256_hex,
    verify_signature,
)
from .registry import Registry, decode_public_key
from .replay import ReplayStore


class VerificationError(ValueError):
    """Structured verifier denial."""

    def __init__(self, code: str, message: str):
        super().__init__(message)
        self.code = code
        self.message = message


@dataclass(frozen=True)
class VerificationResult:
    principal: PrincipalRecord
    scopes: tuple[str, ...]
    evidence: AuthorizationEvidence

    def to_dict(self) -> dict[str, Any]:
        return {
            "principal": self.principal.to_dict(),
            "scopes": list(self.scopes),
            "evidence": self.evidence.to_dict(),
        }


def _parse_registry_time(value: str | None, field: str) -> datetime | None:
    if value is None:
        return None
    try:
        return datetime.fromisoformat(value.replace("Z", "+00:00"))
    except ValueError as exc:
        raise VerificationError("INVALID_REGISTRY", f"{field} is not valid RFC3339") from exc


class Verifier:
    def __init__(
        self,
        registry: Registry,
        replay_store: ReplayStore,
        *,
        expected_audience: str,
        expected_environment: str,
        audit_sink: AuditSink | None = None,
        max_ttl_seconds: int = MAX_TTL_SECONDS,
        clock_skew_seconds: int = 0,
    ) -> None:
        if not expected_audience or not expected_environment:
            raise ValueError("expected audience and environment are required")
        if max_ttl_seconds <= 0 or max_ttl_seconds > MAX_TTL_SECONDS:
            raise ValueError("max_ttl_seconds must be between 1 and 300")
        if clock_skew_seconds < 0:
            raise ValueError("clock_skew_seconds cannot be negative")
        self.registry = registry
        self.replay_store = replay_store
        self.expected_audience = expected_audience
        self.expected_environment = expected_environment
        self.audit_sink = audit_sink
        self.max_ttl_seconds = max_ttl_seconds
        self.clock_skew_seconds = clock_skew_seconds

    def _audit(self, event: AuditEvent) -> None:
        if self.audit_sink is not None:
            self.audit_sink.append(event)

    def _reject(
        self,
        code: str,
        message: str,
        *,
        payload: dict[str, Any] | None = None,
        request: RequestContext | None = None,
        action: str = "verify_agent_request",
    ) -> None:
        safe = payload or {}
        method = ""
        canonical_path = ""
        body_digest = ""
        if request is not None:
            method = request.method.upper() if isinstance(request.method, str) else ""
            try:
                canonical_path = canonical_request_target(request.target)
            except Exception:
                canonical_path = ""
            body_digest = sha256_hex(request.body) if isinstance(request.body, bytes) else ""
        event = AuditEvent(
            event_id=str(uuid.uuid4()),
            principal_type="machine",
            principal_id=str(safe.get("iss", "")) if isinstance(safe.get("iss", ""), str) else "",
            key_id=str(safe.get("kid", "")) if isinstance(safe.get("kid", ""), str) else "",
            environment=str(safe.get("environment", "")) if isinstance(safe.get("environment", ""), str) else "",
            audience=str(safe.get("aud", "")) if isinstance(safe.get("aud", ""), str) else "",
            scope=str(safe.get("scope", "")) if isinstance(safe.get("scope", ""), str) else "",
            action=action,
            http_method=method or str(safe.get("http_method", "")),
            canonical_path=canonical_path or str(safe.get("canonical_path", "")),
            body_sha256=body_digest or str(safe.get("body_sha256", "")),
            resource=safe.get("resource") if isinstance(safe.get("resource"), str) else None,
            request_id=request.request_id or str(safe.get("request_id", "")) if request else str(safe.get("request_id", "")),
            jti=str(safe.get("jti", "")) if isinstance(safe.get("jti", ""), str) else "",
            result="REJECTED",
            result_code=code,
            created_at=datetime.now(timezone.utc).isoformat(timespec="seconds").replace("+00:00", "Z"),
        )
        self._audit(event)
        raise VerificationError(code, message)

    def verify(self, assertion: str, request: RequestContext, *, now: int, action: str = "verify_agent_request") -> VerificationResult:
        payload: dict[str, Any] | None = None
        try:
            parsed = parse_assertion(assertion)
            payload = parsed.payload
        except AssertionErrorCode as exc:
            self._reject(exc.code, exc.message, request=request, action=action)
        assert payload is not None
        try:
            if isinstance(now, bool) or not isinstance(now, int) or now < 0:
                self._reject("INVALID_CLOCK", "verifier now must be a non-negative integer", payload=payload, request=request, action=action)
            if payload["exp"] - payload["iat"] > self.max_ttl_seconds:
                self._reject("TTL_EXCEEDED", "assertion exceeds verifier TTL", payload=payload, request=request, action=action)
            if now + self.clock_skew_seconds < payload["nbf"]:
                self._reject("NOT_YET_VALID", "assertion is not yet valid", payload=payload, request=request, action=action)
            if now - self.clock_skew_seconds >= payload["exp"]:
                self._reject("EXPIRED", "assertion has expired", payload=payload, request=request, action=action)

            key_record = self.registry.keys.get(payload["kid"])
            if key_record is None:
                self._reject("UNKNOWN_KEY", "key is not registered", payload=payload, request=request, action=action)
            if key_record.algorithm != "Ed25519":
                self._reject("UNSUPPORTED_KEY_ALGORITHM", "key algorithm is not Ed25519", payload=payload, request=request, action=action)
            if not key_record.enabled:
                self._reject("KEY_DISABLED", "key is disabled", payload=payload, request=request, action=action)
            if key_record.revoked_at is not None:
                self._reject("KEY_REVOKED", "key is revoked", payload=payload, request=request, action=action)
            expires_at = _parse_registry_time(key_record.expires_at, "key expires_at")
            if expires_at is not None and datetime.now(expires_at.tzinfo) >= expires_at:
                self._reject("KEY_EXPIRED", "key is expired", payload=payload, request=request, action=action)

            principal = self.registry.principals.get(key_record.principal_id)
            if principal is None:
                self._reject("UNKNOWN_PRINCIPAL", "key owner principal is not registered", payload=payload, request=request, action=action)
            if not principal.enabled:
                self._reject("PRINCIPAL_DISABLED", "principal is disabled", payload=payload, request=request, action=action)
            if principal.revoked_at is not None:
                self._reject("PRINCIPAL_REVOKED", "principal is revoked", payload=payload, request=request, action=action)
            if payload["principal_epoch"] != principal.revocation_epoch:
                self._reject("PRINCIPAL_EPOCH_MISMATCH", "principal revocation epoch is stale", payload=payload, request=request, action=action)
            if payload["key_epoch"] != key_record.key_epoch:
                self._reject("KEY_EPOCH_MISMATCH", "key epoch is stale", payload=payload, request=request, action=action)

            public_key = Ed25519PublicKey.from_public_bytes(decode_public_key(key_record.public_key))
            verify_signature(parsed, public_key)

            if payload["iss"] != principal.principal_id or payload["sub"] != principal.principal_id:
                self._reject("WRONG_ISSUER", "assertion issuer is not the registered key owner", payload=payload, request=request, action=action)
            if payload["aud"] != self.expected_audience:
                self._reject("WRONG_AUDIENCE", "assertion audience does not match verifier", payload=payload, request=request, action=action)
            if payload["environment"] != self.expected_environment or payload["environment"] != principal.environment:
                self._reject("WRONG_ENVIRONMENT", "assertion environment does not match verifier", payload=payload, request=request, action=action)
            if payload["http_method"] != request.method.upper():
                self._reject("METHOD_MISMATCH", "HTTP method is not bound by the assertion", payload=payload, request=request, action=action)
            if payload["canonical_path"] != canonical_request_target(request.target):
                self._reject("PATH_MISMATCH", "request target is not bound by the assertion", payload=payload, request=request, action=action)
            if payload["content_type"] != normalize_content_type(request.content_type):
                self._reject("CONTENT_TYPE_MISMATCH", "content type is not bound by the assertion", payload=payload, request=request, action=action)
            if payload["body_sha256"] != sha256_hex(request.body):
                self._reject("BODY_DIGEST_MISMATCH", "request body is not bound by the assertion", payload=payload, request=request, action=action)
            if request.request_id is None or payload["request_id"] != request.request_id:
                self._reject("REQUEST_ID_MISMATCH", "request ID is not bound by the assertion", payload=payload, request=request, action=action)
            if "resource" in payload and payload["resource"] != request.resource:
                self._reject("RESOURCE_MISMATCH", "resource is not bound by the assertion", payload=payload, request=request, action=action)
            if not self.registry.has_grant(
                principal_id=principal.principal_id,
                environment=payload["environment"],
                audience=payload["aud"],
                scope=payload["scope"],
                resource=payload.get("resource"),
            ):
                self._reject("SCOPE_DENIED", "no exact scope grant matches the assertion", payload=payload, request=request, action=action)
            if not self.replay_store.consume(payload["jti"], payload["exp"], now=now):
                self._reject("REPLAYED_JTI", "assertion JTI has already been consumed", payload=payload, request=request, action=action)
        except VerificationError:
            raise
        except AssertionErrorCode as exc:
            self._reject(exc.code, exc.message, payload=payload, request=request, action=action)
        except (ValueError, TypeError) as exc:
            self._reject("INVALID_REGISTRY", str(exc), payload=payload, request=request, action=action)

        evidence = AuthorizationEvidence(
            principal_id=principal.principal_id,
            key_id=key_record.key_id,
            environment=payload["environment"],
            audience=payload["aud"],
            scope=payload["scope"],
            request_id=payload["request_id"],
            jti=payload["jti"],
            canonical_path=payload["canonical_path"],
            body_sha256=payload["body_sha256"],
            resource=payload.get("resource"),
        )
        self._audit(
            AuditEvent(
                event_id=str(uuid.uuid4()),
                principal_type="machine",
                principal_id=principal.principal_id,
                key_id=key_record.key_id,
                environment=payload["environment"],
                audience=payload["aud"],
                scope=payload["scope"],
                action=action,
                http_method=payload["http_method"],
                canonical_path=payload["canonical_path"],
                body_sha256=payload["body_sha256"],
                resource=payload.get("resource"),
                request_id=payload["request_id"],
                jti=payload["jti"],
                result="AUTHORIZED",
                result_code="AUTHORIZED",
                created_at=datetime.now(timezone.utc).isoformat(timespec="seconds").replace("+00:00", "Z"),
            )
        )
        return VerificationResult(principal=principal, scopes=(payload["scope"],), evidence=evidence)

File: typescript/src/verifier.ts
import { createPublicKey, randomUUID } from "node:crypto";

import {
  AssertionPayload,
  MAX_TTL_SECONDS,
  ProtocolError,
  canonicalRequestTarget,
  normalizeContentType,
  parseAssertion,
  sha256Hex,
  verifySignature,
} from "./protocol.js";

export interface PrincipalRecord {
  principal_id: string;
  display_name: string;
  environment: string;
  enabled: boolean;
  revoked_at?: string | null;
  revocation_epoch: number;
  created_at?: string;
  updated_at?: string;
}

export interface KeyRecord {
  key_id: string;
  principal_id: string;
  algorithm: string;
  public_key: string;
  enabled: boolean;
  expires_at?: string | null;
  revoked_at?: string | null;
  key_epoch: number;
  created_at?: string;
}

export interface ScopeGrant {
  principal_id: string;
  environment: string;
  audience: string;
  scope: string;
  resource?: string | null;
}

export interface Registry {
  principals: PrincipalRecord[];
  keys: KeyRecord[];
  grants: ScopeGrant[];
}

export interface RequestContext {
  method: string;
  target: string;
  body: Uint8Array;
  contentType?: string;
  requestId?: string;
  resource?: string;
}

export interface ReplayStore {
  consume(jti: string, expiresAt: number, now: number): boolean;
}

export class MemoryReplayStore implements ReplayStore {
  private readonly entries = new Map<string, number>();

  consume(jti: string, expiresAt: number, now: number): boolean {
    for (const [key, expiry] of this.entries) if (expiry <= now) this.entries.delete(key);
    if (this.entries.has(jti)) return false;
    this.entries.set(jti, expiresAt);
    return true;
  }
}

export interface AuditEvent {
  event_id: string;
  principal_type: "machine";
  principal_id: string;
  key_id: string;
  environment: string;
  audience: string;
  scope: string;
  action: string;
  http_method: string;
  canonical_path: string;
  body_sha256: string;
  resource: string | null;
  request_id: string;
  jti: string;
  result: "AUTHORIZED" | "REJECTED";
  result_code: string;
  created_at: string;
}

export interface AuditSink {
  append(event: AuditEvent): void;
}

export class VerificationError extends Error {
  readonly code: string;

  constructor(code: string, message: string) {
    super(message);
    this.name = "VerificationError";
    this.code = code;
  }
}

export interface AuthorizationEvidence {
  principal_id: string;
  key_id: string;
  environment: string;
  audience: string;
  scope: string;
  request_id: string;
  jti: string;
  canonical_path: string;
  body_sha256: string;
  resource: string | null;
}

export interface VerificationResult {
  principal: PrincipalRecord;
  scopes: string[];
  evidence: AuthorizationEvidence;
}

function reject(code: string, message: string): never {
  throw new VerificationError(code, message);
}

function decodePublicKey(value: string): Buffer {
  if (!/^[A-Za-z0-9_-]+$/u.test(value)) reject("INVALID_REGISTRY", "public key is not valid base64url");
  const raw = Buffer.from(value, "base64url");
  if (raw.toString("base64url") !== value || raw.length !== 32) reject("INVALID_REGISTRY", "Ed25519 public keys must be 32 bytes");
  return raw;
}

function publicKeyObject(value: string): ReturnType<typeof createPublicKey> {
  const spkiPrefix = Buffer.from("302a300506032b6570032100", "hex");
  return createPublicKey({ key: Buffer.concat([spkiPrefix, decodePublicKey(value)]), format: "der", type: "spki" });
}

function parseExpiry(value: string | null | undefined): number | null {
  if (value === undefined || value === null) return null;
  const parsed = Date.parse(value);
  if (Number.isNaN(parsed)) reject("INVALID_REGISTRY", "key expires_at is not valid RFC3339");
  return Math.floor(parsed / 1000);
}

function findPrincipal(registry: Registry, id: string): PrincipalRecord | undefined {
  return registry.principals.find((item) => item.principal_id === id);
}

function findKey(registry: Registry, id: string): KeyRecord | undefined {
  return registry.keys.find((item) => item.key_id === id);
}

function hasGrant(registry: Registry, payload: AssertionPayload): boolean {
  return registry.grants.some(
    (grant) =>
      grant.principal_id === payload.iss &&
      grant.environment === payload.environment &&
      grant.audience === payload.aud &&
      grant.scope === payload.scope &&
      (grant.resource === undefined || grant.resource === null || grant.resource === payload.resource),
  );
}

function isoNow(): string {
  return new Date().toISOString();
}

export function verifyAgentRequest(options: {
  assertion: string;
  request: RequestContext;
  registry: Registry;
  replayStore: ReplayStore;
  expectedAudience: string;
  expectedEnvironment: string;
  now: number;
  action?: string;
  auditSink?: AuditSink;
  maxTtlSeconds?: number;
  clockSkewSeconds?: number;
}): VerificationResult {
  let payload: AssertionPayload | undefined;
  const action = options.action ?? "verify_agent_request";
  const audit = (result: "AUTHORIZED" | "REJECTED", code: string, value: AssertionPayload | undefined): void => {
    const requestPath = (() => {
      try {
        return canonicalRequestTarget(options.request.target);
      } catch {
        return "";
      }
    })();
    options.auditSink?.append({
      event_id: randomUUID(),
      principal_type: "machine",
      principal_id: value?.iss ?? "",
      key_id: value?.kid ?? "",
      environment: value?.environment ?? "",
      audience: value?.aud ?? "",
      scope: value?.scope ?? "",
      action,
      http_method: typeof options.request.method === "string" ? options.request.method.toUpperCase() : "",
      canonical_path: requestPath || value?.canonical_path || "",
      body_sha256: sha256Hex(options.request.body),
      resource: value?.resource ?? null,
      request_id: options.request.requestId ?? value?.request_id ?? "",
      jti: value?.jti ?? "",
      result,
      result_code: code,
      created_at: isoNow(),
    });
  };
  const deny = (code: string, message: string): never => {
    audit("REJECTED", code, payload);
    return reject(code, message);
  };
  try {
    payload = parseAssertion(options.assertion).payload;
  } catch (error) {
    const protocol = error as ProtocolError;
    audit("REJECTED", protocol.code ?? "MALFORMED_ASSERTION", undefined);
    throw new VerificationError(protocol.code ?? "MALFORMED_ASSERTION", protocol.message);
  }
  const maxTtl = options.maxTtlSeconds ?? MAX_TTL_SECONDS;
  const skew = options.clockSkewSeconds ?? 0;
  if (typeof options.request.method !== "string" || options.request.method.length === 0) deny("INVALID_REQUEST", "HTTP method is required");
  if (!Number.isSafeInteger(options.now) || options.now < 0) deny("INVALID_CLOCK", "verifier now must be a non-negative integer");
  if (maxTtl <= 0 || maxTtl > MAX_TTL_SECONDS || skew < 0) deny("INVALID_CONFIGURATION", "invalid verifier time configuration");
  if (payload.exp - payload.iat > maxTtl) deny("TTL_EXCEEDED", "assertion exceeds verifier TTL");
  if (options.now + skew < payload.nbf) deny("NOT_YET_VALID", "assertion is not yet valid");
  if (options.now - skew >= payload.exp) deny("EXPIRED", "assertion has expired");

  const key = findKey(options.registry, payload.kid);
  if (!key) {
    audit("REJECTED", "UNKNOWN_KEY", payload);
    throw new VerificationError("UNKNOWN_KEY", "key is not registered");
  }
  if (key.algorithm !== "Ed25519") deny("UNSUPPORTED_KEY_ALGORITHM", "key algorithm is not Ed25519");
  if (!key.enabled) deny("KEY_DISABLED", "key is disabled");
  if (key.revoked_at) deny("KEY_REVOKED", "key is revoked");
  const expiry = parseExpiry(key.expires_at);
  if (expiry !== null && options.now >= expiry) deny("KEY_EXPIRED", "key is expired");
  const principal = findPrincipal(options.registry, key.principal_id);
  if (!principal) {
    audit("REJECTED", "UNKNOWN_PRINCIPAL", payload);
    throw new VerificationError("UNKNOWN_PRINCIPAL", "key owner principal is not registered");
  }
  if (!principal.enabled) deny("PRINCIPAL_DISABLED", "principal is disabled");
  if (principal.revoked_at) deny("PRINCIPAL_REVOKED", "principal is revoked");
  if (payload.principal_epoch !== principal.revocation_epoch) deny("PRINCIPAL_EPOCH_MISMATCH", "principal revocation epoch is stale");
  if (payload.key_epoch !== key.key_epoch) deny("KEY_EPOCH_MISMATCH", "key epoch is stale");
  try {
    verifySignature(parseAssertion(options.assertion), publicKeyObject(key.public_key));
  } catch (error) {
    const protocol = error as ProtocolError;
    deny(protocol.code ?? "BAD_SIGNATURE", protocol.message);
  }
  if (payload.iss !== principal.principal_id || payload.sub !== principal.principal_id) deny("WRONG_ISSUER", "assertion issuer is not the registered key owner");
  if (payload.aud !== options.expectedAudience) deny("WRONG_AUDIENCE", "assertion audience does not match verifier");
  if (payload.environment !== options.expectedEnvironment || payload.environment !== principal.environment) deny("WRONG_ENVIRONMENT", "assertion environment does not match verifier");
  if (payload.http_method !== options.request.method.toUpperCase()) deny("METHOD_MISMATCH", "HTTP method is not bound by the assertion");
  if (payload.canonical_path !== canonicalRequestTarget(options.request.target)) deny("PATH_MISMATCH", "request target is not bound by the assertion");
  if (payload.content_type !== normalizeContentType(options.request.contentType)) deny("CONTENT_TYPE_MISMATCH", "content type is not bound by the assertion");
  if (payload.body_sha256 !== sha256Hex(options.request.body)) deny("BODY_DIGEST_MISMATCH", "request body is not bound by the assertion");
  if (options.request.requestId === undefined || payload.request_id !== options.request.requestId) deny("REQUEST_ID_MISMATCH", "request ID is not bound by the assertion");
  if (payload.resource !== undefined && payload.resource !== options.request.resource) deny("RESOURCE_MISMATCH", "resource is not bound by the assertion");
  if (!hasGrant(options.registry, payload)) deny("SCOPE_DENIED", "no exact scope grant matches the assertion");
  if (!options.replayStore.consume(payload.jti, payload.exp, options.now)) deny("REPLAYED_JTI", "assertion JTI has already been consumed");
  const evidence: AuthorizationEvidence = {
    principal_id: principal.principal_id,
    key_id: key.key_id,
    environment: payload.environment,
    audience: payload.aud,
    scope: payload.scope,
    request_id: payload.request_id,
    jti: payload.jti,
    canonical_path: payload.canonical_path,
    body_sha256: payload.body_sha256,
    resource: payload.resource ?? null,
  };
  audit("AUTHORIZED", "AUTHORIZED", payload);
  return { principal, scopes: [payload.scope], evidence };
}
